File: SomethingObviouslyUseless/Analyzers/SelfAssignment/SelfAssignmentAnalyzer/SelfAssignmentAnalyzer/SelfAssignmentAnalyzer.cs
using System.Collections.Immutable;
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.Diagnostics;
using Microsoft.CodeAnalysis.Operations;

namespace SelfAssignmentAnalyzer
{
    [DiagnosticAnalyzer(LanguageNames.CSharp)]
    public class SelfAssignmentAnalyzer : DiagnosticAnalyzer
    {
        private const string Category = "SelfAssignment";

        private static readonly DiagnosticDescriptor SimpleSelfAssignmentRule = new DiagnosticDescriptor("SACS0001", "Self assignment detected by C#.", "Redundant self assignment detected by C#.", Category, DiagnosticSeverity.Warning, true);
        private static readonly DiagnosticDescriptor PropertySelfReferenceRule = new DiagnosticDescriptor("SACS0002", "Property self reference detected by C#.", "Property self reference detected by C#. Potentially a bug.", Category, DiagnosticSeverity.Warning, true);
        private static readonly DiagnosticDescriptor PropertySelfCompoundAssignmentRule = new DiagnosticDescriptor("SACS0003", "Property self compound assignment detected by C#.", "Property self compound assignment detected by C#. Change this to a method", Category, DiagnosticSeverity.Warning, true);

        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics =>
            ImmutableArray.Create(SimpleSelfAssignmentRule, PropertySelfReferenceRule, PropertySelfCompoundAssignmentRule);

        public override void Initialize(AnalysisContext context)
        {
            context.RegisterOperationAction(AnalyzeSimpleAssignment, OperationKind.SimpleAssignment);
            context.RegisterOperationAction(AnalyzeCompoundAssignment, OperationKind.CompoundAssignment);
            context.RegisterOperationAction(AnalyzePropertyReference, OperationKind.PropertyReference);
        }

        private static bool IsMatch<T>(IOperation target, IOperation value)
            => target is T tTarget && value is T tValue && tTarget.Equals(tValue);
        
        private static void AnalyzeSimpleAssignment(OperationAnalysisContext context)
        {
            var operation = (ISimpleAssignmentOperation)context.Operation;

            if (IsMatch<IParameterReferenceOperation>(operation.Target, operation.Value)
             || IsMatch<IPropertyReferenceOperation>(operation.Target, operation.Value)
             || IsMatch<IFieldReferenceOperation>(operation.Target, operation.Value))
            {
                var diagnostic = Diagnostic.Create(SimpleSelfAssignmentRule, operation.Syntax.GetLocation());
                context.ReportDiagnostic(diagnostic);
            }
        }

        private static void AnalyzeCompoundAssignment(OperationAnalysisContext context)
        {
            var operation = (ICompoundAssignmentOperation)context.Operation;

            if (IsMatch<IParameterReferenceOperation>(operation.Target, operation.Value))
            {
                var diagnostic = Diagnostic.Create(PropertySelfCompoundAssignmentRule, operation.Syntax.GetLocation());
                context.ReportDiagnostic(diagnostic);
            }
        }

        private static void AnalyzePropertyReference(OperationAnalysisContext context)
        {
            var operation = (IPropertyReferenceOperation)context.Operation;

            if (context.ContainingSymbol is IMethodSymbol methodSymbol
                && methodSymbol.AssociatedSymbol is IPropertySymbol property
                && operation.Property.Equals(property))
            {
                var diagnostic = Diagnostic.Create(PropertySelfReferenceRule, operation.Syntax.GetLocation());
                context.ReportDiagnostic(diagnostic);
            }
        }
    }
}

File: SomethingObviouslyUseless/Analyzers/SelfAssignment/SelfAssignmentAnalyzerFSharp/SelfAssignmentAnalyzer.fs
namespace SelfAssignmentAnalyzerFSharp

open Microsoft.CodeAnalysis.Diagnostics
open Microsoft.CodeAnalysis
open System.Collections.Immutable
open Microsoft.CodeAnalysis.Operations

module DiagnosticCreator =
    let createDiagnostic (context: OperationAnalysisContext) (descriptor: DiagnosticDescriptor) =
        Diagnostic.Create(descriptor, context.Operation.Syntax.GetLocation())
        |> context.ReportDiagnostic

module Operation =
    let equalsAs<'a when 'a :> IOperation and 'a: equality> (target: IOperation) (value: IOperation) =
        match target, value with
        | :? 'a as typedTarget, (:? 'a as typedValue)
            when typedTarget = typedValue -> true
        | _ -> false

module SimpleAssignmentAnalysis =
    let rule = DiagnosticDescriptor("SAFS0001", "Self assignment detected by F#", "Redundant self assignment detected by F#.", "SelfAssignment F#",
                                    DiagnosticSeverity.Warning, true, null, null, array.Empty<string>())        
    let analyze (context : OperationAnalysisContext) =
        let operation = context.Operation :?> ISimpleAssignmentOperation
        
        match operation.Target, operation.Value with
        | target, value when target |> Operation.equalsAs<IPropertyReferenceOperation> value -> DiagnosticCreator.createDiagnostic context rule
        | target, value when target |> Operation.equalsAs<IParameterReferenceOperation> value -> DiagnosticCreator.createDiagnostic context rule
        | target, value when target |> Operation.equalsAs<IFieldReferenceOperation> value ->  DiagnosticCreator.createDiagnostic context rule
        | _ -> ()

module PropertyReferenceAnalysis =
    let rule = DiagnosticDescriptor("SAFS0002", "Property self reference detected by F#", "Property self reference detected by F#. Potentially a bug.", "SelfAssignment F#",
                                    DiagnosticSeverity.Warning, true, null, null, array.Empty<string>())
    let analyze (context : OperationAnalysisContext) =
        let operation = context.Operation :?> IPropertyReferenceOperation
    
        match context.ContainingSymbol with
        | :? IMethodSymbol as methodSymbol
            when methodSymbol.AssociatedSymbol.Equals(operation.Property, SymbolEqualityComparer.Default) -> DiagnosticCreator.createDiagnostic context rule
        | _ -> ()

module CompoundAssignmentAnalysis =
    let rule =  DiagnosticDescriptor("SAFS0003", "Property self compound assignment detected by F#", "Property self compound assignment detected by F#. Change this to a method.", "SelfAssignment F#",
                                     DiagnosticSeverity.Warning, true, null, null, array.Empty<string>());        
    let analyze (context : OperationAnalysisContext) =
        let operation = context.Operation :?> ICompoundAssignmentOperation
    
        match operation.Target, operation.Value with
        | target, value when target |> Operation.equalsAs<IPropertyReferenceOperation> value -> DiagnosticCreator.createDiagnostic context rule
        | _ -> ()

[<DiagnosticAnalyzer(LanguageNames.CSharp)>]
type SelfAssignmentAnalyzerFSharp() =
    inherit DiagnosticAnalyzer()

    override this.SupportedDiagnostics =
               ImmutableArray.Create(SimpleAssignmentAnalysis.rule, CompoundAssignmentAnalysis.rule, PropertyReferenceAnalysis.rule)   

    override this.Initialize (context : AnalysisContext) =
        context.EnableConcurrentExecution();
        context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze ||| GeneratedCodeAnalysisFlags.ReportDiagnostics);
        context.RegisterOperationAction(SimpleAssignmentAnalysis.analyze, OperationKind.SimpleAssignment)
        context.RegisterOperationAction(CompoundAssignmentAnalysis.analyze, OperationKind.CompoundAssignment)
        context.RegisterOperationAction(PropertyReferenceAnalysis.analyze, OperationKind.PropertyReference)